### PR TITLE
NMS-12385: Handle exclude-range in Discovery definitions

### PR DESCRIPTION
**Pull request by cgorantla**
`exclude-range` now supports location.
 If `exclude-range` is in definitions it inherits location from definition.

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-12385
* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-12386
* Bamboo (Continuous Integration): https://bamboo.opennms.org/

### CODE REVIEW

File: opennms-config/src/main/java/org/opennms/netmgt/config/DiscoveryConfigFactory.java [hunk 1]
```diff
@@ -525,18 +525,29 @@ private void clearUrlSpecifics() {
...
     @Override
-    public boolean isExcluded(final InetAddress address) {
+    public boolean isExcluded(final InetAddress address, String location) {
         getReadLock().lock();
         try {
-            final List<ExcludeRange> excludeRange = getConfiguration().getExcludeRanges();
-            if (excludeRange != null) {
+            final List<ExcludeRange> excludeRanges = getConfiguration().getExcludeRanges();
+            // Get exclude ranges from definitions. Assign location from definition.
+            getConfiguration().getDefinitions().forEach(def -> {
```
Comment (j-white) [inline opennms-config/src/main/java/org/opennms/netmgt/config/DiscoveryConfigFactory.java]: Shouldn't this be performed when the configuration is loaded/reloaded as opposed to being done on every call to `isExcluded`?

File: opennms-config-model/src/main/java/org/opennms/netmgt/config/discovery/ExcludeRange.java [hunk 2]
```diff
@@ -98,14 +116,20 @@ public boolean equals(final Object obj) {
...
+                    && Objects.equals(temp.m_end, m_end)
+                    && Objects.equals(temp.m_location, m_location);
         }
         return false;
     }
 
     @Override
     public String toString() {
-        return "ExcludeRange [begin=" + m_begin + ", end=" + m_end + "]";
+        return "ExcludeRange{" +
```
Comment (j-white) [inline opennms-config-model/src/main/java/org/opennms/netmgt/config/discovery/ExcludeRange.java]: We normally don't include the `m_` prefix in the toString labels. While we're changing this, consider using `MoreObjects.ToStringHelper`.

File: opennms-config-model/src/main/java/org/opennms/netmgt/config/discovery/ExcludeRange.java [hunk 3]
```diff
@@ -58,6 +59,14 @@
     @XmlElement(name = "end", required = true)
     private String m_end;
 
+    /**
+     * The monitoring location where this exclude range
+     *  will be excluded
+     */
+    @XmlAttribute(name = "location")
```
Comment (j-white) [inline opennms-config-model/src/main/java/org/opennms/netmgt/config/discovery/ExcludeRange.java]: We should add support for this field to the UI as well, since we already support configuring the exclude ranges there.

**Comment by cgorantla** [comment]
discovery-exclude-include-location

Handled review comments  and updated UI with location field in `exclude-range`

**Comment by cgorantla** [comment]
I attempted to fix ProvisionerTest but looks like it's already taken care in foundation-2019. Merging foundation-2019 should fix integration tests.

File: opennms-webapp/src/main/webapp/admin/discovery/add-ir.jsp [hunk 4]
```diff
@@ -14,7 +14,7 @@
  *
  * OpenNMS(R) is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
```
Comment (j-white) [inline opennms-webapp/src/main/webapp/admin/discovery/add-ir.jsp]: Unnecessary changes in this file.

File: opennms-webapp/src/main/webapp/admin/discovery/add-er.jsp [hunk 5]
```diff
@@ -29,7 +29,17 @@
 
 --%>
 <%@page language="java" contentType="text/html" session="true" import="
-  org.opennms.web.admin.discovery.DiscoveryServletConstants
+  java.util.Map,
+  java.util.TreeMap,
+  org.opennms.netmgt.model.monitoringLocations.OnmsMonitoringLocation,
+  org.opennms.netmgt.dao.api.*,
```
Comment (j-white) [inline opennms-webapp/src/main/webapp/admin/discovery/add-er.jsp]: Avoid * imports.

File: opennms-webapp/src/main/java/org/opennms/web/admin/discovery/ActionDiscoveryServlet.java [hunk 6]
```diff
@@ -244,9 +244,14 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
             LOG.debug("Adding Exclude Range");
             String ipAddrBegin = request.getParameter("erbegin");
             String ipAddrEnd = request.getParameter("erend");
+            String location = request.getParameter("erlocation");
             ExcludeRange newER = new ExcludeRange();
             newER.setBegin(ipAddrBegin);
             newER.setEnd(ipAddrEnd);
+            if(location!=null && !"".equals(location.trim()) &&
```
Comment (j-white) [inline opennms-webapp/src/main/java/org/opennms/web/admin/discovery/ActionDiscoveryServlet.java]: Can we simplify, or explain this conditional? Would be nice to add a utility function to `LocationUtils` for this type of check.
